src/main/java/com/learning/mfscreener/service/CapitalGainsService.java

```diff
@@ -0,0 +1,145 @@
+
+    private static final double MIN_OPEN_BALANCE = 0.01;
+
+    private BigDecimal investedAmount = BigDecimal.ZERO;
+    private Double currentValue = 0D;
+    private final List<GainEntry> gainEntries = new ArrayList<>();
+    List<String> errors = new ArrayList<>();
+
+    @Loggable(params = false)
+    Map<String, Object> processData(CasDTO casDTO) throws IncompleteCASError {
+        casDTO.folios().forEach(this::processFolio);
+        return prepareGains(gainEntries);
+    }
+
+    void processFolio(UserFolioDTO folio) {
+        folio.schemes().forEach(scheme -> processScheme(folio, scheme));
+    }
+
+    void processScheme(UserFolioDTO folio, UserSchemeDTO scheme) {
+        List<UserTransactionDTO> transactions = scheme.transactions();
+        Fund fund = new Fund(scheme.scheme(), folio.folio(), scheme.isin(), scheme.type());
+        if (!transactions.isEmpty()) {
+            validateOpenBalance(scheme);
+            processTransactions(fund, transactions, scheme);
+        }
+    }
+
+    void validateOpenBalance(UserSchemeDTO scheme) {
+        if (Double.parseDouble(scheme.myopen()) >= MIN_OPEN_BALANCE) {
+            throw new IncompleteCASError(
+                    "Incomplete CAS found. For gains computation all folios should have zero opening balance");
+        }
+    }
+
+    void processTransactions(Fund fund, List<UserTransactionDTO> transactions, UserSchemeDTO scheme) {
+        try {
+            FIFOUnits fifo = new FIFOUnits(fund, transactions);
+            investedAmount = this.investedAmount.add(fifo.getInvested());
+            currentValue += scheme.valuation().value();
+            gainEntries.addAll(fifo.getGains());
+        } catch (GainsException exc) {
+            this.errors.add(fund.scheme() + ", " + exc.getMessage());
+        }
+    }
+
+    Map<String, Object> prepareGains(List<GainEntry> gainEntries) {
+        Map<String, Object> summary = new HashMap<>();
+        // Sort the gains by fy and fund
+        gainEntries.sort(Comparator.comparing(GainEntry::getFinYear).thenComparing(GainEntry::getFundType));
+        // Group the gains by fy and fund
+        Map<String, List<GainEntry>> groupedGains = new HashMap<>();
+        for (GainEntry txn : gainEntries) {
+            String key = txn.getFinYear() + "-" + txn.getFundType();
+            if (!groupedGains.containsKey(key)) {
+                groupedGains.put(key, new ArrayList<>());
+            }
+            groupedGains.get(key).add(txn);
+        }
+        for (Map.Entry<String, List<GainEntry>> entry : groupedGains.entrySet()) {
+            String[] keys = entry.getKey().split("-");
+            String fy = keys[0];
+            FundType fund = FundType.valueOf(keys[1]);
+            if (!summary.containsKey(fy)) {
+                summary.put(fy, new HashMap<String, Object>() {
+                    {
+                        put("funds", new ArrayList<>());
+                        put("total", new HashMap<String, BigDecimal>() {
```

Comment:
@coderabbitai, Use another way to initialize this instance. Double Brace Initialization should not be used [java:S3599](https://sonarcloud.io/organizations/rajadilipkolli/rules?open=java%3AS3599&rule_key=java%3AS3599)